internal/resource/config/datastore/ldap_data_store.go

```diff
@@ -0,0 +1,566 @@
+						Required:    true,
+						ElementType: types.StringType,
+					},
+					"tags": schema.StringAttribute{
+						Description: "Tags associated with this data source.",
+						Optional:    true,
+					},
+					"default_source": schema.BoolAttribute{
+						Description: "Whether this is the default connection. Defaults to false if not specified.",
+						Default:     booldefault.StaticBool(false),
+						Computed:    true,
+						Optional:    true,
+					},
+				},
+			},
+			PlanModifiers: []planmodifier.Set{
+				setplanmodifier.UseStateForUnknown(),
+			},
+			Validators: []validator.Set{
+				setvalidator.AtLeastOneOf(
+					path.Expression.AtName(path.MatchRoot("ldap_data_store"), "hostnames_tags"),
+					path.Expression.AtName(path.MatchRoot("ldap_data_store"), "hostnames"),
+				),
+			},
+		},
+		"time_between_evictions": schema.Int64Attribute{
+			Description: "The frequency, in milliseconds, that the evictor cleans up the connections in the pool. A value of -1 disables the evictor. Omitting this attribute will set the value to the default value.",
+			Computed:    true,
+			Optional:    true,
+			PlanModifiers: []planmodifier.Int64{
+				int64planmodifier.UseStateForUnknown(),
+			},
+		},
+		"user_dn": schema.StringAttribute{
+			Description: "The username credential required to access the data store.",
+			Required:    true,
+		},
+		"password": schema.StringAttribute{
+			Description: "The password credential required to access the data store. GETs will not return this attribute. To update this field, specify the new value in this attribute.",
+			Required:    true,
+			Sensitive:   true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.UseStateForUnknown(),
+			},
+		},
+		"bind_anonymously": schema.BoolAttribute{
+			Description: "Whether username and password are required. The default value is false.",
+			Computed:    true,
+			Optional:    true,
+			Default:     booldefault.StaticBool(false),
+		},
+		"follow_ldap_referrals": schema.BoolAttribute{
+			Description: "Follow LDAP Referrals in the domain tree. The default value is false. This property does not apply to PingDirectory as this functionality is configured in PingDirectory.",
+			Computed:    true,
+			Optional:    true,
+			Default:     booldefault.StaticBool(false),
+		},
+	}
+
+	ldapDataStoreSchema.Validators = []validator.Object{
+		objectvalidator.ExactlyOneOf(
+			path.MatchRelative().AtParent().AtName("custom_data_store"),
+			path.MatchRelative().AtParent().AtName("jdbc_data_store"),
+			path.MatchRelative().AtParent().AtName("ping_one_ldap_gateway_data_store"),
+		),
+	}
+
+	return ldapDataStoreSchema
+}
+
+func toStateLdapDataStore(con context.Context, clientValue *client.DataStoreAggregation, plan basetypes.ObjectValue) (types.Object, diag.Diagnostics) {
+	var diags, allDiags diag.Diagnostics
+	ldapDataStore := *clientValue.LdapDataStore
+
+	userDn := func() types.String {
+		if *ldapDataStore.BindAnonymously {
```

Comment:
Probably need a nil check here

Reply:
nil checks are now added to all data store types

Reply:
It looks like BindAnonymously itself could be nil too here. And in the nil check above for ldapDataStore, we should return there, or we could still have a panic here.

Reply:
This line has been changed to below
`if ldapDataStore.BindAnonymously != nil && *ldapDataStore.BindAnonymously`